### ISSUE TITLE
Revogi Strip provides Milliwatts

### ISSUE THREAD

**Issue by lmaertin** [Superkikim/mh-maxsmart-hass#4]
The Revogi Strip provides the consumption in Milliwatts. Here /config/custom_components/maxsmart/sensor.py has to be adapted. My quickfix:

`self._power_data = float(power_data['watt']/1000.0)`

It seems sensor.py needs a switch case for the product variant.

**Comment by Superkikim** [comment]
Great tx. I've been advised by another user of the same issue :) I'll fix it in 2023.8.0-beta2.

**Comment by Superkikim** [comment]
Can you share the result of command GET 511 ? 

https://github.com/Superkikim/mh-maxsmart-powerstation/blob/master/cmd-511.md

**Comment by lmaertin** [comment]
Thanks for your engagement in this improvement... here is an examplary output auf the 511 cmd
```json
{"response":511,"code":200,"data":{"watt":[106258,2927,0,0,0,6186],"amp":[475,0,0,140,0,58],"switch":[1,1,0,1,0,1]}}
```

**Comment by Superkikim** [comment]
Damn these guys are stupid. They are providing milliwatts but they kept the label watt ! Seriously. 

I was thinking on how I can differentiate except on firmware version. Because I don't know the fw version that exists for the different OEM providers of this product. 

Maxsmart has ver 1.30 (local metadata) and 2.11 (cloud metadata). I see yours is Revogi with 3.x (cloud). 

Right now, the only check I make is on version 1.30. If not version 1.30, then I set watt from milliwatt and I set generic port names. 

Publishing right now on dev. Will be done in a minute.

**Comment by lmaertin** [comment]
yes, the code quality is as bad as the security of this devices...

amp = Ampere? This is also nonsense with these numbers...

**Comment by Superkikim** [comment]
Please try the 2023.8.0-beta2 version published on dev branch.

**Comment by lmaertin** [comment]
Ports are identified and translation works as well, but the POWER entities are gone now...

**Comment by lmaertin** [comment]
File "/config/custom_components/maxsmart/sensor.py", line 87
    if self._device_version =! "1.30":
                            ^
SyntaxError: invalid syntax

I think this is the root cause

**Comment by lmaertin** [comment]
yes it is... typo in the not equals... with "!=" all works fine...

**Comment by Superkikim** [comment]
> yes it is... typo in the not equals... with "!=" all works fine...

Stupid mistake :laughing: 

Fixed with latest commit, retagged with same version

**Comment by Superkikim** [comment]
By the way @lmaertin, looking at this, I wondered: 

`{"response":511,"code":200,"data":{"watt":[106258,2927,0,0,0,6186],"amp":[475,0,0,140,0,58],"switch":[1,1,0,1,0,1]}}`

looking at this, is this right ? you have a device on slot 1 consuming at over 1000W ?

**Comment by lmaertin** [comment]
it's 106 Watts...

I found another bug in the division by 1000 in the dev branch. Please adapt the code as follows:

```
def update(self, now=None):
    _LOGGER.debug("Entering update")
    power_data = self._maxsmart_device.get_power_data(self._port_id)
    _LOGGER.debug("power_data has been populated with %s",power_data)
    if power_data is not None:
        # Check if the device version is 1.30 and convert from milliwatts to watts if true
        _LOGGER.debug("Firmware version is %s",self._device_version)
        self._power_data = float(power_data['watt'])
        if self._device_version != "1.30":
            self._power_data /= 1000.0
    else:
        self._power_data = 0
```

**Comment by Superkikim** [comment]
Thank you for your support Lukas. I did it the wrong way. trying to make calculation on a string. 

Anyway, it should be fixed. I have updated tag and release to the same version for now. 

Please redownload from HACS. Please enable "Show beta versions". I have published it as pre-release this time to be consistent.

**Comment by lmaertin** [comment]
Works! Thank you so much

**Comment by Superkikim** [comment]
> Works! Thank you so much

Hey Lukas. Still have the REVOGI strips ? Could you share the raw answer from the udp call ? I'm currently refactoring the module (https://github.com/Superkikim/maxsmart) and trying to find a way to decide automatically if it should calculate the watts from milliwatts in that module rather than in the integration. 

I see two leads: 

1. Most accurate: "It if's REVOGI and version 3.??" (Need the version), then it's milliwatts
2. If it's over 1000, it's milliwatts. 

Second, I'm not a fan, because I wonder if an outlet could potentially provide 0.5W for example in some occasions.

**Comment by tmtom** [comment]
Would it be possible to distinguish based on the JSON string whether it looks like float or int (whether it has decimal dot like `1.23` or not like `1230`)? https://github.com/Superkikim/mh-maxsmart-powerstation/blob/master/cmd-511.md suggest Watt to be as float? Can't confirm though as my power strips reports milliwatts and milliamps.

Btw I have Revogi SOW323, FW 3.49 and Cocoso SOW323 with FW 1.06 and both use ints (miliwatts/miliamps) like:
`{"response":511,"code":200,"data":{"watt":[10710,0,0,0,0,0],"amp":[102,0,0,0,0,0],"switch":[1,0,0,0,0,0]}}`

The Cocoso seems to be just rebranded Revogi where for some reason they use different FW numbering (it is this one https://gzhls.at/blob/ldb/6/4/4/f/cd460345576de703ee13e6b0cd983f17beea.pdf)

I did not get what do you need to get tested but if you explain I can try.

**Comment by Superkikim** [comment]
> Would it be possible to distinguish based on the JSON string whether it looks like float or int (whether it has decimal dot like `1.23` or not like `1230`)? https://github.com/Superkikim/mh-maxsmart-powerstation/blob/master/cmd-511.md suggest Watt to be as float? Can't confirm though as my power strips reports milliwatts and milliamps.
> 
> Btw I have Revogi SOW323, FW 3.49 and Cocoso SOW323 with FW 1.06 and both use ints (miliwatts/miliamps) like: `{"response":511,"code":200,"data":{"watt":[10710,0,0,0,0,0],"amp":[102,0,0,0,0,0],"switch":[1,0,0,0,0,0]}}`
> 
> The Cocoso seems to be just rebranded Revogi where for some reason they use different FW numbering (it is this one https://gzhls.at/blob/ldb/6/4/4/f/cd460345576de703ee13e6b0cd983f17beea.pdf)
> 
> I did not get what do you need to get tested but if you explain I can try.

Hey. I got back on work on this. 

I'm currently refactoring the python module to make it better, more reliable, with robust testing, and data detection. I took your advice to detect float/int rather than hard code versions in the module. 

I'll release it hopefully in the coming days. 

When done, I will get back to the Home assistant integration which deserves refactoring and a better module ;-)

**Comment by Superkikim** [comment]
So, to summarize: 

Revogi and Ccoso SOW323, FW 3.49 and fw 1.06 get mW as int

Revogi with 3.36 also get mW as int

Maxsmart SOP111 FW 1.10 and 1.30 get W as float, and 2.11 get mW as int 

@lmaertin , @tmtom , May I as you to test this ?

Create a venv, and test the show_consumption script. 

```
python3 -m venv maxsmart_test
source maxsmart_test/bin/activate
```

If you use Windows, use this to activate: 
```
maxsmart_test\Scripts\activate
```

Then, install the maxsmart 2.0.3 module: 

```
pip3 install maxsmart
```

And run this test script. 

[show_consumption.py](https://github.com/Superkikim/maxsmart/blob/main/example_scripts/show_consumption.py)

Save it in the same folder, or anywhere you want, then: 

```
chmod +x show_consumption.py
./show_consumption.py
```

It will discover the devices, and you can select a device to get the realtime consumption. 

Please post the result here.

**Comment by lmaertin** [comment]
Thanks for the instructions. I can test it in one week, as I have no local access to the installation right now. I will post results than.

**Comment by tmtom** [comment]
In my case the reported power seems to be 1000x higher, the real consumption was about 2-3 Watts.

```shell
 ./show_consumption.py
MaxSmart Device Consumption Monitor
===================================
Monitor real-time power consumption with auto-detection

🔍 Scanning network for MaxSmart devices...
Discovering MaxSmart devices...
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 1/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 2/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 3/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 4/4)
WARNING:root:Missing translation for key 'ERROR_UNEXPECTED' in language 'en' - using default
ERROR:root:MaxSmart log [LOG_STATE_CHANGE] - Missing parameter 'port' for message formatting (available: old_state=unknown, new_state=error, ip=192.168.2.200, detail=Unexpected error getting hardware identifiers: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 4/4))
✅ Found 1 device(s)

Available MaxSmart devices:
1. Revogi black - 192.168.2.200 (FW: 3.49)
  R. Rescan network
  0. Exit
Select device number, R to rescan, or 0 to exit: 1

Connecting to Revogi black (192.168.2.200)...

============================================================
DEVICE INFORMATION
============================================================
Name: None
IP Address: 192.168.2.200
Serial Number: None
Firmware Version: None
Data Format: int_watt
Conversion Factor: 1.0
Total Consumption: 2185.00W
Active Ports: 1/6
Device Type: 6-Port Power Station
============================================================
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 1/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 2/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 3/4)
ERROR:root:Connection error for device 192.168.2.200: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 4/4)
WARNING:root:Missing translation for key 'ERROR_UNEXPECTED' in language 'en' - using default
ERROR:root:MaxSmart log [LOG_STATE_CHANGE] - Missing parameter 'port' for message formatting (available: old_state=unknown, new_state=error, ip=192.168.2.200, detail=Unexpected error getting hardware identifiers: Connection refused by device 192.168.2.200: Connection failed: ServerDisconnectedError - Server disconnected (attempt 4/4))

Real-time Port Consumption:
Port Name | Watts  | State
--------------------------
Port 1    | 0.00   | OFF
Port 2    | 0.00   | OFF
Port 3    | 2332.00 | ON
Port 4    | 0.00   | OFF
Port 5    | 0.00   | OFF
Port 6    | 0.00   | OFF

Summary:
  Total Consumption: 2332.00W
  Active Ports: 1/6
  Device Type: 6-Port Power Station
  Average per Active Port: 2332.00W

Options:
  1. Refresh data (same device)
  2. Select different device
  3. Exit

Select option (1-3): 3

Goodbye!
```

**Comment by Superkikim** [comment]
Damn🤬. There seem indeed to be a bug.

```
Data Format: int_watt
Conversion Factor: 1.0
```

If it's `int` factor should be `0.001`. 

I'll check that today. 

Also I'm surprised by the number of retries. Do you have any specific problem on your network? Perhaps I should increase the time out on tries.

**Comment by Superkikim** [comment]
Regarding the connection errors, Claude has implemented keepalive which is most probably refused by the devices, and it doesn't make sense anyway.

I'll prepare a new release most probably tonight for the maxsmart module and then script. If you can test it it would be great. My devices are less restrictive. 

I'd be glad if you can send me you phone nr at superkikim@sissaoui.com so we can chat on Telegram or WhatsApp and we can expedite the matter. Let me just know when it's a good time for you.

**Comment by Superkikim** [comment]
> In my case the reported power seems to be 1000x higher, the real consumption was about 2-3 Watts.
> 

Could you please try again and give me the output ? 

You should already have the venv, so you just need to reactivated, and reinstall maxsmart

`pip install maxsmart` or `pip3 install maxsmart"

You should receive maxsmart 2.0.4

**Comment by tmtom** [comment]
Seems OK now:
```shell
...
Firmware Version: 3.49
Data Format: int_milliwatt
Conversion Factor: 0.001
Total Consumption: 2.09W
Active Ports: 1/6
Device Type: 6-Port Power Station
============================================================

Real-time Port Consumption:
Port Name | Watts | State
-------------------------
Port 1    | 0.00  | OFF
Port 2    | 0.00  | OFF
Port 3    | 2.09  | ON
Port 4    | 0.00  | OFF
Port 5    | 0.00  | OFF
Port 6    | 0.00  | OFF

Summary:
  Total Consumption: 2.09W
  Active Ports: 1/6
  Device Type: 6-Port Power Station
  Average per Active Port: 2.09W
...
```

**Comment by Superkikim** [comment]
> Seems OK now

Nice. I'm fixing some stuff in the integration, then I will release 2025.8.1

You should be informed by HACS as soon as it is available.

**Comment by tmtom** [comment]
FYI I have pushed some notes from my analysis how the Revogi cloud setup works - see https://github.com/tmtom/RevogiPowerstripProtocol 
Still incomplete but beleieve it would be possible to create own setup/provisioning app and own "cloud server" and let the powerstrip connect and talk to it instead. I.e get rid off both the Revogi phone app and the server.
Otherwise the only advantage might be slightly more efficient communication, question if worth the effort...

**Comment by Superkikim** [comment]
Fixed in 2025.8.1. Please update from HACS

**Comment by lmaertin** [comment]
Works for me, thanks a lot!
